### ISSUE TITLE
 Unrecognized unit 'degrees_north' for axis 'rlat'

### ISSUE THREAD

**Issue by rkouznetsov** [Unidata/tds#286]
TDS 5.5-SNAPSHOT. A request
`http://thredds.silam.fmi.fi/thredds/ncss/grid/silam_europe_pollen_v5_8/runs/silam_europe_pollen_v5_8_RUN_2022-09-18T00:00:00Z?var=cnc_POLLEN_RAGWEED_m18&spatial=bb&north=56&west=4&east=16&south=46&temporal=range&time_start=2022-09-19T00:00:00Z&time_end=2022-09-19T00:00:00Z&accept=netcdf3`
pefectly works and produces a reasonable file (except for the `_NCProperties` attribute, see https://github.com/Unidata/tds/issues/284), however  the `threddsServlet.log` file gets filled with hundreds thousands of messages:
```
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_east' for axis 'rlon'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_north' for axis 'rlat'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_east' for axis 'rlon'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_north' for axis 'rlat'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_east' for axis 'rlon'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_north' for axis 'rlat'
2022-09-19T09:00:01.498 +0300 [  76037016][   15625] INFO  - ucar.nc2.ft2.coverage.HorizCoordSys - Unrecognized unit 'degrees_east' for axis 'rlon'
```
So there are two issues here: 
1. Can the reporting be less wordy?
2. What makes the perfectly CF-compliant units unrecognized?

Thank you!

**Comment by tdrwenski** [comment]
Thanks for reporting this. I recently add this error message but it was definitely not the intention that it would produce so many messages! I will have a look.

**Comment by ethanrd** [comment]
Hi @rkouznetsov - A quick CF response to your question 2.

While `degrees_north` and `degrees_east` are valid lat/lon units, their use is not recommended for the [rotated pole grid mapping](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#_rotated_pole). This bit on units is described in CF sections [4.1 "Latitude Coordinate"](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#latitude-coordinate) and [4.2 "Longitude Coordinate"](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#longitude-coordinate) (last paragraphs). From 4.1 (similar for longitude in 4.2):

>Coordinates of latitude with respect to a rotated pole should be given units of degrees, not degrees_north or equivalents, because applications which use the units to identify axes would have no means of distinguishing such an axis from real latitude, and might draw incorrect coastlines, for instance.

The text uses a "should" not a "must" so I expect the [CF Checker](https://pumatest.nerc.ac.uk/cgi-bin/cf-checker.pl) would give a warning but not an error. But I haven't tested it.

But how the TDS handles it is a different matter. TDS (and netCDF-Java) tries to be lenient with CF rules when reading from existing files. So, as @tdrwenski mentions, it probably shouldn't complain at all.

**Comment by rkouznetsov** [comment]
Thank you, @ethanrd! I have changed the rotated-coordinate units in Silam -dev branch accordingly. Hushing this warning is worth it anyway...

I have a vague memories that `_CoordinateAxisType` was introduced long ago, just because of thredds, who did not recognize grid parameters otherwise. Now NetCDF folks decided to claim all `_underscored` attributes for internal use of the library. Are these underscored attributes in use in current TDS?

Is there any single document that describes how TDS (or netcdfJava) recognizes coordinates, axes, verticals etc?. Some years ago I have spent quite some time to guess which attributes must be set to make thredds do the right job on our grids...

**Comment by tdrwenski** [comment]
I have removed the log message-- it was indeed complaining about things it shouldn't have been and logging too much as well.

**Comment by ethanrd** [comment]
@rkouznetsov - The netCDF-C library has long claimed underscore attributes (e.g., `_FillValue`) as reserved for the library. It is spelled out in the [NetCDF User's Guide (NUG)](https://docs.unidata.ucar.edu/nug/current/index.html) in [Appendix A](https://docs.unidata.ucar.edu/netcdf-c/current/attribute_conventions.html). The NUG is undergoing a reorganization and one change will likely be to change that language from "reserved for use by the netCDF library" to something like "reserved for use by netCDF implementations".

The [`_Coordinate` attributes](https://docs.unidata.ucar.edu/netcdf-java/current/userguide/coord_attr_conv.html) are supported by the netCDF-Java library so are somewhat covered by the netCDF underscore attribute rule. Though they really extend beyond the netCDF data model and implementations and into conventions for representing geospatial information. I believe that these `_Coordinate` attributes helped inform the [CF Conventions](https://cfconventions.org) `coordinates` and `axis` attributes described in CF [Section 5 "Coordinate Systems and Domain"](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#coordinate-system).

Bottom line: We recommend following the CF Conventions instead of the `_Coordinate` attributes, if at all possible.

**Comment by rkouznetsov** [comment]
Thank you, @ethanrd for the clarifications!  I'll consider getting rid of `_Coordinate`,  `_CoordinateModelRunDate` etc. What would be the right method to communicate the simulation start time to thredds? Currently we have both `_CoordinateModelRunDate` and `SIMULATION_START_DATE`.  The latter has been monkeyed from WRF, but, to my surprise, has not been yet pushed to CF.

CF conventions have an example of  a separate variable in addition to the usual time axis:
```
double atime
    atime:standard_name = "forecast_reference_time" ;
    atime:units = "hours since 1999-01-01 00:00" ;
```
Is it the same as `_CoordinateModelRunDate`, and can TDS interpret and handle those?
Can that be not the first time step of a forecast? We have a few suites that start from a day or two ago to catchup with the observations that take long to deliver. Every time we have to explain to a new user that the forecast started from two days ago is actually the today's one...

**Comment by rkouznetsov** [comment]
Thank you for the fix, @tdrwenski!  Will it be enough to recompile TDS with `./gradlew war` on the main branch of TDS to get the fix, or is there a special trick needed to bring updated netcdf-java into .war?

**Comment by ethanrd** [comment]
@rkouznetsov - Yes, the "forecast_reference_time" standard name is how to indicate the start time in a CF compliant manner. You would also reference your "atime" variable in a "coordinates" attribute on each of your data variables, e.g., `var1:coordinates = "atime time lat lon`.

Hope that helps. If you have more questions about using CF, please create a separate issue so we don't distract too much from the focus of this issue.

**Comment by rkouznetsov** [comment]
Thank you, @tdrwenski,
I have download and deployed the snapshot you suggested, but the issue is still there. Luckily thredds has a mechanism that prevents it from generating more than 100MB/hour logs :).
I have re-started tomcat, but it did not help either..
After  removing my previous `thredds-5.5-SNAPSHOT.war` and making `git fetch && git rebase && ./gradlew --refresh-dependencies && ./gradlew war ` on the TDS main branch  the messages seem to have gone. A new issue appeared though: https://github.com/Unidata/tds/issues/293

**Comment by rkouznetsov** [comment]
Thank you, @ethanrd, very helpful. I'll make new issues then..

**Comment by tdrwenski** [comment]
Sorry about that link to the jar not being one with the fix, but glad the gradle command worked. We will have a look at your new issues.

**Comment by rkouznetsov** [comment]
No worries. Thank you, @tdrwenski!  Without your `refresh-dependencies` magic it would need few more iterations :)
Looking forward to finally decommission 4.5 and switch to 5.x!
